# Importing Libraries.

In [1]:
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

# Setting Path.

In [2]:
path = r'C:\Users\Chaot\02-2024 Instacart Basket Analysis'
path

'C:\\Users\\Chaot\\02-2024 Instacart Basket Analysis'

# 4.9.3 - Import your analysis libraries, as well as your new customer data set as a dataframe.

In [4]:
op_merged = pd.read_pickle(os.path.join(path, '02 - Data', 'Prepared Data', '4.8_agg_orders_products_merge.pkl'))

In [5]:
op_merged.head(10)

Unnamed: 0.1  Unnamed: 0_x  order_id  user_id  order_number  \
0             0             0   2539329        1             1   
1             1             1   2398795        1             2   
2             2             2    473747        1             3   
3             3             3   2254736        1             4   
4             4             4    431534        1             5   
5             5             5   3367565        1             6   
6             6             6    550135        1             7   
7             7             7   3108588        1             8   
8             8             8   2295261        1             9   
9             9             9   2550362        1            10   

   orders_day_of_week  order_hour_of_day  days_since_prior_order  product_id  \
0                   2                  8                     NaN         196   
1                   3                  7                    15.0         196   
2                   3                 12                    21.0         196   
3                   4                  7                    29.0         196   
4                   4                 15                    28.0         196   
5                   2                  7                    19.0         196   
6                   1                  9                    20.0         196   
7                   1                 14                    14.0         196   
8                   1                 16                     0.0         196   
9                   4                  8                    30.0         196   

   add_to_cart_order  ...       price_range_loc     busiest_day  \
0                  1  ...  Medium Range Product  Regularly busy   
1                  1  ...  Medium Range Product  Regularly busy   
2                  1  ...  Medium Range Product  Regularly busy   
3                  1  ...  Medium Range Product      Least busy   
4                  1  ...  Medium Range Product      Least busy   
5                  1  ...  Medium Range Product  Regularly busy   
6                  1  ...  Medium Range Product  Regularly busy   
7                  2  ...  Medium Range Product  Regularly busy   
8                  4  ...  Medium Range Product  Regularly busy   
9                  1  ...  Medium Range Product      Least busy   

          busiest_days busiest_period_of_day  max_order  loyalty_flag  \
0  Regularly busy days        Average Orders         10  New Customer   
1   Least busiest days        Average Orders         10  New Customer   
2   Least busiest days           Most Orders         10  New Customer   
3   Least busiest days        Average Orders         10  New Customer   
4   Least busiest days           Most Orders         10  New Customer   
5  Regularly busy days        Average Orders         10  New Customer   
6         Busiest days           Most Orders         10  New Customer   
7         Busiest days           Most Orders         10  New Customer   
8         Busiest days           Most Orders         10  New Customer   
9   Least busiest days        Average Orders         10  New Customer   

   mean_product_price spender_flag Customer_frequency frequency_flag  
0            6.367797          Low               20.5   Non-Frequent  
1            6.367797          Low               20.5   Non-Frequent  
2            6.367797          Low               20.5   Non-Frequent  
3            6.367797          Low               20.5   Non-Frequent  
4            6.367797          Low               20.5   Non-Frequent  
5            6.367797          Low               20.5   Non-Frequent  
6            6.367797          Low               20.5   Non-Frequent  
7            6.367797          Low               20.5   Non-Frequent  
8            6.367797          Low               20.5   Non-Frequent  
9            6.367797          Low               20.5   Non-Frequent  

[10 rows x 28 columns]

In [34]:
df_customer = pd.read_csv(os.path.join(path, '02 - Data', 'Original Data', 'customers.csv'), index_col = False)

In [35]:
df_customer

user_id First Name    Surnam  Gender           STATE  Age date_joined  \
0         26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1         33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2         65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3        125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4        130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
...         ...        ...       ...     ...             ...  ...         ...   
206204   168073       Lisa      Case  Female  North Carolina   44    4/1/2020   
206205    49635     Jeremy   Robbins    Male          Hawaii   62    4/1/2020   
206206   135902      Doris  Richmond  Female        Missouri   66    4/1/2020   
206207    81095       Rose   Rollins  Female      California   27    4/1/2020   
206208    80148    Cynthia     Noble  Female        New York   55    4/1/2020   

        n_dependants fam_status  income  
0                  3    married  165665  
1                  0     single   59285  
2                  2    married   99568  
3                  0     single   42049  
4                  1    married   40374  
...              ...        ...     ...  
206204             1    married  148828  
206205             3    married  168639  
206206             2    married   53374  
206207             1    married   99799  
206208             1    married   57095  

[206209 rows x 10 columns]

# 4.9.4 - Wrangle the data so that it follows consistent logic; for example, rename columns with illogical names and drop columns that don’t add anything to your analysis.

# 4.9.5 - Complete the fundamental data quality and consistency checks you’ve learned throughout this Achievement; for example, check for and address missing values and duplicates, and convert any mixed-type data.

In [36]:
#Initial check of column names.

df_customer.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [41]:
#Fixing column names for customer dataframe.

df_customer.rename(columns = {'First Name' : 'first_name',
                              'Surnam' : 'last_name',
                              'Gender' : 'gender',
                              'STATE' : 'state',
                              'Age' : 'age'}, inplace = True)

In [44]:
#Check to find nulls and datatype.

df_customer.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  int64 
 1   first_name    194950 non-null  object
 2   last_name     206209 non-null  object
 3   gender        206209 non-null  object
 4   state         206209 non-null  object
 5   age           206209 non-null  int64 
 6   date_joined   206209 non-null  object
 7   n_dependants  206209 non-null  int64 
 8   fam_status    206209 non-null  object
 9   income        206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


In [46]:
#Check to find nulls.

df_customer['first_name'].value_counts(dropna = False)

first_name
NaN        11259
Marilyn     2213
Barbara     2154
Todd        2113
Jeremy      2104
           ...  
Merry        197
Eugene       197
Garry        191
Ned          186
David        186
Name: count, Length: 208, dtype: int64

In [51]:
#Check to find nulls.

df_customer.isnull().sum()

user_id             0
first_name      11259
last_name           0
gender              0
state               0
age                 0
date_joined         0
n_dependants        0
fam_status          0
income              0
dtype: int64

Although we've identified 11,259 missing first names, these may not be needed since they all have a user ID and last name. For now we'll move on to finding if there are duplicates.

In [52]:
df_cust_dupe = df_customer[df_customer.duplicated()]

In [53]:
df_cust_dupe

Empty DataFrame
Columns: [user_id, first_name, last_name, gender, state, age, date_joined, n_dependants, fam_status, income]
Index: []

There are no duplicate values.

We will now move on to the op_merged dataframe.

In [49]:
#Checking for columns + datatype.
op_merged.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404859 entries, 0 to 32404858
Data columns (total 28 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   Unnamed: 0.1            int64   
 1   Unnamed: 0_x            int64   
 2   order_id                int64   
 3   user_id                 int64   
 4   order_number            int64   
 5   orders_day_of_week      int64   
 6   order_hour_of_day       int64   
 7   days_since_prior_order  float64 
 8   product_id              int64   
 9   add_to_cart_order       int64   
 10  reordered               int64   
 11  _merge                  category
 12  Unnamed: 0_y            int64   
 13  product_name            object  
 14  aisle_id                int64   
 15  department_id           int64   
 16  prices                  float64 
 17  indicator2              category
 18  price_range_loc         object  
 19  busiest_day             object  
 20  busiest_days            object  
 21  busies

In [50]:
#dropping columns for op_merged check.

op_merged.drop(columns = ['Unnamed: 0.1', 'Unnamed: 0_x', '_merge', 'indicator2', 'Unnamed: 0_y', 'busiest_day'])

order_id  user_id  order_number  orders_day_of_week  \
0          2539329        1             1                   2   
1          2398795        1             2                   3   
2           473747        1             3                   3   
3          2254736        1             4                   4   
4           431534        1             5                   4   
...            ...      ...           ...                 ...   
32404854   1320836   202557            17                   2   
32404855     31526   202557            18                   5   
32404856    758936   203436             1                   2   
32404857   2745165   203436             2                   3   
32404858   3093936   205420             1                   4   

          order_hour_of_day  days_since_prior_order  product_id  \
0                         8                     NaN         196   
1                         7                    15.0         196   
2                        12                    21.0         196   
3                         7                    29.0         196   
4                        15                    28.0         196   
...                     ...                     ...         ...   
32404854                 15                     1.0       43553   
32404855                 11                     3.0       43553   
32404856                  7                     NaN       42338   
32404857                  5                    15.0       42338   
32404858                 14                     NaN       28818   

          add_to_cart_order  reordered                   product_name  ...  \
0                         1          0                           Soda  ...   
1                         1          1                           Soda  ...   
2                         1          1                           Soda  ...   
3                         1          1                           Soda  ...   
4                         1          1                           Soda  ...   
...                     ...        ...                            ...  ...   
32404854                  2          1            Orange Energy Shots  ...   
32404855                  2          1            Orange Energy Shots  ...   
32404856                  4          0          Zucchini Chips, Pesto  ...   
32404857                 16          1          Zucchini Chips, Pesto  ...   
32404858                  8          0  Hot Oatmeal Multigrain Raisin  ...   

          prices       price_range_loc         busiest_days  \
0            9.0  Medium Range Product  Regularly busy days   
1            9.0  Medium Range Product   Least busiest days   
2            9.0  Medium Range Product   Least busiest days   
3            9.0  Medium Range Product   Least busiest days   
4            9.0  Medium Range Product   Least busiest days   
...          ...                   ...                  ...   
32404854     3.7     Low Range Product  Regularly busy days   
32404855     3.7     Low Range Product  Regularly busy days   
32404856     6.9  Medium Range Product  Regularly busy days   
32404857     6.9  Medium Range Product   Least busiest days   
32404858    10.3  Medium Range Product   Least busiest days   

         busiest_period_of_day max_order      loyalty_flag  \
0               Average Orders        10      New Customer   
1               Average Orders        10      New Customer   
2                  Most Orders        10      New Customer   
3               Average Orders        10      New Customer   
4                  Most Orders        10      New Customer   
...                        ...       ...               ...   
32404854           Most Orders        31  Regular Customer   
32404855           Most Orders        31  Regular Customer   
32404856        Average Orders         3      New Customer   
32404857         Fewest Orders         3      New Customer   
32404858           Most Orders        16  Regula

# 4.9.6 - Combine your customer data with the rest of your prepared Instacart data. (Hint: Make sure the key columns are the same data type!)

In [54]:
# Merging (C)ustomers (O)rders and (P)roducts together.

cop_merged = op_merged.merge(df_customer, on = 'user_id')

In [59]:
#Check to see if they merged.
cop_merged.head(25)

Unnamed: 0.1  Unnamed: 0_x  order_id  user_id  order_number  \
0              0             0   2539329        1             1   
1              1             1   2398795        1             2   
2              2             2    473747        1             3   
3              3             3   2254736        1             4   
4              4             4    431534        1             5   
5              5             5   3367565        1             6   
6              6             6    550135        1             7   
7              7             7   3108588        1             8   
8              8             8   2295261        1             9   
9              9             9   2550362        1            10   
10             0             0   2539329        1             1   
11             0             0   2539329        1             1   
12             1             1   2398795        1             2   
13             2             2    473747        1             3   
14             3             3   2254736        1             4   
15             4             4    431534        1             5   
16             5             5   3367565        1             6   
17             6             6    550135        1             7   
18             7             7   3108588        1             8   
19             8             8   2295261        1             9   
20             9             9   2550362        1            10   
21             0             0   2539329        1             1   
22             1             1   2398795        1             2   
23             0             0   2539329        1             1   
24             3             3   2254736        1             4   

    orders_day_of_week  order_hour_of_day  days_since_prior_order  product_id  \
0                    2                  8                     NaN         196   
1                    3                  7                    15.0         196   
2                    3                 12                    21.0         196   
3                    4                  7                    29.0         196   
4                    4                 15                    28.0         196   
5                    2                  7                    19.0         196   
6                    1                  9                    20.0         196   
7                    1                 14                    14.0         196   
8                    1                 16                     0.0         196   
9                    4                  8                    30.0         196   
10                   2                  8                     NaN       14084   
11                   2                  8                     NaN       12427   
12                   3                  7                    15.0       12427   
13                   3                 12                    21.0       12427   
14                   4                  7                    29.0       12427   
15                   4                 15                    28.0       12427   
16                   2                  7                    19.0       12427   
17                   1                  9                    20.0       12427   
18                   1                 14                    14.0       12427   
19                   1                 16                     0.0       12427   
20                   4                  8                    30.0       12427   
21                   2                  8                     NaN       26088   
22                   3                  7                    15.0       26088   
23                   2                  8                     NaN       26405   
24                   4                  7                    29.0       26405   

    add_to_cart_order  ...  frequency_flag first_name  last_name  gender  \
0                   1  ...    Non-Frequent      Linda     Nguyen  Female   
1 

In [61]:
#Column name and data type check.

cop_merged.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404859 entries, 0 to 32404858
Data columns (total 37 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   Unnamed: 0.1            int64   
 1   Unnamed: 0_x            int64   
 2   order_id                int64   
 3   user_id                 int64   
 4   order_number            int64   
 5   orders_day_of_week      int64   
 6   order_hour_of_day       int64   
 7   days_since_prior_order  float64 
 8   product_id              int64   
 9   add_to_cart_order       int64   
 10  reordered               int64   
 11  _merge                  category
 12  Unnamed: 0_y            int64   
 13  product_name            object  
 14  aisle_id                int64   
 15  department_id           int64   
 16  prices                  float64 
 17  indicator2              category
 18  price_range_loc         object  
 19  busiest_day             object  
 20  busiest_days            object  
 21  busies

In [63]:
#Realized that I didn't save the drop columns from the op_merged so I'm doing it again.

cop_merged_dc = cop_merged.drop(columns = ['Unnamed: 0.1', 'Unnamed: 0_x', '_merge', 'Unnamed: 0_y', 'indicator2', 'busiest_day'])

In [65]:
#Unifying column title header.

cop_merged_dc.rename(columns = {'Customer_frequency' : 'customer_frequency'}, inplace = True)

In [67]:
#Checking columns to make sure its updated.
cop_merged_dc.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404859 entries, 0 to 32404858
Data columns (total 31 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                int64  
 1   user_id                 int64  
 2   order_number            int64  
 3   orders_day_of_week      int64  
 4   order_hour_of_day       int64  
 5   days_since_prior_order  float64
 6   product_id              int64  
 7   add_to_cart_order       int64  
 8   reordered               int64  
 9   product_name            object 
 10  aisle_id                int64  
 11  department_id           int64  
 12  prices                  float64
 13  price_range_loc         object 
 14  busiest_days            object 
 15  busiest_period_of_day   object 
 16  max_order               int64  
 17  loyalty_flag            object 
 18  mean_product_price      float64
 19  spender_flag            object 
 20  customer_frequency      float64
 21  frequency_flag          objec

# 4.9.8 - Export this new dataframe as a pickle file so you can continue to use it in the second part of this task.

In [70]:
cop_merged_dc.to_pickle(os.path.join(path, '02 - Data', 'Prepared Data', '4.9_cop_merged.pkl'))